src/aspire/storage/starfile.py

```diff
@@ -1,173 +1,181 @@
-                        tokens = tokens[: len(field_names)]  # ignore any extra tokens
-
-                    rows.append(tokens)
-
-            # Any pending rows to be added?
-            if rows:
-                loops.append(pd.DataFrame(rows, columns=field_names, dtype=str))
-
-            # Any pending loops/properties to be added?
-            if loops or properties:
-                blocks.append(
-                    StarFileBlock(loops, name=block_name, properties=properties)
-                )
+                if gemmi_item.loop is not None:
+                    block_has_loop = True
+                    # if we find both a pair and a loop raise an error
+                    if block_has_pair:
+                        raise StarFileError(
+                            "Blocks with multiple loops and/or pairs are not supported"
+                        )
+                    loop_tags = gemmi_item.loop.tags
+                    # convert loop data to a list of lists
+                    # using the .val(row, col) method of gemmi's Loop class
+                    loop_data = [None] * gemmi_item.loop.length()
+                    for row in range(gemmi_item.loop.length()):
+                        loop_data[row] = [
+                            gemmi_item.loop.val(row, col)
+                            for col in range(gemmi_item.loop.width())
+                        ]
+            if block_has_pair:
+                if gemmi_block.name not in self.blocks:
+                    # initialize DF from dictionary of kv pairs
+                    self.blocks[gemmi_block.name] = pd.DataFrame(
+                        [pairs], columns=pairs.keys(), dtype=str
+                    )
+                else:
+                    # enforce unique keys
+                    raise StarFileError(
+                        f"Attempted overwrite of existing data block: {gemmi_block.name}"
+                    )
+            elif block_has_loop:
+                if gemmi_block.name not in self.blocks:
+                    # initialize DF from list of lists
+                    self.blocks[gemmi_block.name] = pd.DataFrame(
+                        loop_data, columns=loop_tags, dtype=str
+                    )
+                else:
+                    # enforce unique keys
+                    raise StarFileError(
+                        f"Attempted overwrite of existing data block: {gemmi_block.name}"
+                    )
 
-            logger.info("StarFile parse complete")
+    def write(self, filepath):
+        """
+        Converts `blocks` to a gemmi Document and writes to a starfile at the given filepath.
+        """
+        # create an empty Document
+        _doc = cif.Document()
+        filepath = str(filepath)
+        for name, df in self.blocks.items():
+            # construct new empty block
+            _block = _doc.add_new_block(name)
+            # are we constructing a pair (df has 1 row) or a loop (df has >1 rows)
```

Comment:
This is an issue. Ideally, we should be able to distinguish between a set of pairs and a loop with one row (these are distinct in the STAR files themselves, so we'd rail the read–write test for those). Again, this is a problem related to our way of representing STAR files internally (with dataframes), not with this PR. It would be good to create an issue here so that we can come back to it.

Reply:
Created #463